### ISSUE TITLE
References to classes generated by other build-generated types are resolved to dynamic (Moor, SuperEnum, BuiltValue)

### ISSUE THREAD

**Issue by shyndman** [mobxjs/mobx.dart#405]
`mobx_codegen` resolves a moor generated class to `dynamic`. I suspect this is a limitation in the `build` framework, and how the analyzer is set up to resolve against generated code.

![image](https://user-images.githubusercontent.com/32758755/72907621-e772fd80-3d12-11ea-81dd-c1e66e3c34e9.png)

![image](https://user-images.githubusercontent.com/32758755/72920161-8570c300-3d27-11ea-9e7e-60fa688c6abc.png)

Versions:
```
flutter_mobx 0.3.6
mobx 0.4.0+1
mobx_codegen 0.4.0+1
build_runner 1.7.3
```

_Originally posted by @irvine5k in https://github.com/mobxjs/mobx.dart/issues/381#issuecomment-577308858_

**Comment by shyndman** [comment]
Similar issue with `built_value`:
https://github.com/simolus3/moor/issues/89#issuecomment-515793437

**Comment by shyndman** [comment]
And the bug in `build`: https://github.com/dart-lang/build/issues/2389

**Comment by SHRJamal** [comment]
This is happening too when using SuperEnum.

**Comment by shyndman** [comment]
Upvoting [this issue](https://github.com/dart-lang/build/issues/2389) is your best bet.

**Comment by shyndman** [comment]
A possible workaround is to create an empty subclass of the SuperEnum/Moor/BuiltValue class, and reference that in your Store class.

**Comment by SHRJamal** [comment]
@shyndman Thanks, I did upvote.
I am not exactly what do you mean by creating an empty subclass, in my case with SuperEnum, what I do now is just change dynamic to the correct Type on every build, which is painfull 😩

**Comment by shyndman** [comment]
I haven't tried this myself, but I recall seeing this kind of thing work previously. Here's what you might do for SuperEnum:

```dart
import 'package:super_enum/super_enum.dart';

part "result.g.dart";

/// You can give this a better name, but notice that it subclasses the class generated by SuperEnum
class MobxVisibleResult extends Result {}

@superEnum
enum _Result {
  // ...
}
```

Then reference `MobxVisibleResult` in your MobX store class, and 🤞 🙏

**Comment by shyndman** [comment]
I wrote a response, and realized it will not work.

Let me think on this for a bit. I do have another solution in mind, but requires changes to the code generator.

**Comment by SHRJamal** [comment]
This sample trick did work for me.
what do you think @shyndman ?

`class EmptyClass{}

class AuthState = AuthStateImp with EmptyClass;

@superEnum
enum _AuthStateImp {
  @object
  AuthInitial,

  @object
  AuthLoading,

  @object
  Authenticated,

  @object
  Unauthenticated
}`

**Comment by irvine5k** [comment]
@shyndman I don't know if this bug has a concrete relationship with this issue but take a look at this: https://github.com/dart-lang/sdk/issues/40428 

A kind of class construction declaration are resolving to dynamic while the same class constructed with different syntax resolves to the correct type.

**Comment by SHRJamal** [comment]
> This sample trick did work for me.
> what do you think @shyndman ?
> 
> ```dart
> class EmptyClass{}
> 
> class AuthState = AuthStateImp with EmptyClass;
> 
> @superEnum
> enum _AuthStateImp {
>   @object
>   AuthInitial,
> ...
> ```
> ```

Oops! This doesn't compile 😰, you can discover it only after you run the code, wonder why the compiler doesn't show any error before then? 🤔🤔

**Comment by elbeshenrique** [comment]
Any updates on this matter?

**Comment by shyndman** [comment]
> @shyndman I don't know if this bug has a concrete relationship with this issue but take a look at this: [dart-lang/sdk#40428](https://github.com/dart-lang/sdk/issues/40428)
> 
> A kind of class construction declaration are resolving to dynamic while the same class constructed with different syntax resolves to the correct type.

No, it's unrelated. It really is the build bug.

No update. It's not an easy one to fix, because it rests solely on how the build system works. We can maybe hack around it, but that has potential to introduce other issues. I'm going to suggest something on the build bug that might help.

**Comment by shyndman** [comment]
I filed a new issue that will hopefully suggest a path forward.

**Comment by shyndman** [comment]
OK, so that was a no-go, but I have another thought. I'm going to play around for a few minutes and see if I can come up with something.

**Comment by shyndman** [comment]
For `SuperEnum`, once https://github.com/xsahil03x/super_enum/pull/32 is released on pub we have a path forward.

The idea would be that you define each of your enum value types as a custom class with `@UseClass`, and because that custom class is not generated, our builder will be able to reference it.

The only other reasonable workaround I can see is having our generator *fallback* on using the exact strings it finds written in the store definition whenever the analyzer tells us a type is dynamic. We used a system like this previously, and it had its drawbacks as the lone driver of type names (did not handle inference of a variable's type), but as a fallback mechanism it might be alright. I need to run it by @pavanpodila

**Comment by SHRJamal** [comment]
> For `SuperEnum`, once [xsahil03x/super_enum#32](https://github.com/xsahil03x/super_enum/pull/32) is released on pub we have a path forward.
> 
> The idea would be that you define each of your enum value types as a custom class with `@UseClass`, and because that custom class is not generated, our builder will be able to see it (ie, not generate `dynamic`)
> 
> The only other reasonable workaround I can see is having our generator _fallback_ on using the exact strings it finds written in the store definition whenever the analyzer tells us a type is dynamic. We used a system like this previously, and it had its drawbacks as the lone driver of type names (did not handle inference of a variable's type), but as a fallback mechanism it might be alright.
> 
> @pavanpodila, what do you think?

About `SuperEnum` the main problem is not about enum values generated, it's more about the enum equivalent Class.

**Comment by shyndman** [comment]
Ah, right, of course.

**Comment by shyndman** [comment]
I wonder whether the new typedefs would help with this? Acting as an equivalent of a forward declaration.

**Comment by SHRJamal** [comment]
> I wonder whether the new typedefs would help with this? Acting as an equivalent of a forward declaration.

New typedefs? 🤔, what I know is that they work only on functions

**Comment by shyndman** [comment]
Yeah, that's the new part — supporting non-function types. https://github.com/dart-lang/language/issues/65

**Comment by SHRJamal** [comment]
> Yeah, that's the new part — supporting non-function types. [dart-lang/language#65](https://github.com/dart-lang/language/issues/65)

Interesting 😍,  I think this will work.

**Comment by shyndman** [comment]
Just looked at the analyzer code. The experiment flag exists, but it isn't hooked up to anything yet. 😞 

OK, let's sit on this. See what Pavan thinks about bringing back the old type namer as a fallback mechanism.

**Comment by shyndman** [comment]
Partial classes will also help for this kind of problem, someday: https://github.com/dart-lang/language/issues/678

**Comment by rrousselGit** [comment]
For what it's worth, [freezed](https://github.com/rrousselgit/freezed) doesn't suffer from that issue.\
That's because it does **not** rely on `SharedPartBuilder`. 


Mobx could fix this by generating a `*.mobx.dart` instead of `*.g.dart`

**Comment by SHRJamal** [comment]
I was using SuperEnum before I discover Freezed 3days ago. It saved that issue for me.
Thanks @rrousselGit for all your awesome Packages.

**Comment by shyndman** [comment]
> For what it's worth, [freezed](https://github.com/rrousselgit/freezed) doesn't suffer from that issue.
> That's because it does **not** rely on `SharedPartBuilder`.
> 
> Mobx could fix this by generating a `*.mobx.dart` instead of `*.g.dart`

Would we have to configure our builder to run after other libraries' builders, in order to guarantee that we see the types?

**Comment by rrousselGit** [comment]
I do not know enough about that.

But what's certain is that even if mobx would need to say "run after XX library", it is configurable from anywhere. 

Simply exposing the `Generator` should be enough for users of the package to fix it themselves.

**Comment by rrousselGit** [comment]
Something like that:

```yaml
# build.yaml
targets:
  $default:
    builders:
      mobx_codegen|mobx_generator:
        enabled: false

builders:
  my_mobx_generator:
    import: "package:example/builder.dart"
    builder_factories: ["mobxBuilder"]
    build_extensions: { ".dart": [".mobx.dart"] }
    auto_apply: dependents
    build_to: source
    required_inputs: [built_value_generator|built_value]
```



```dart
// example/lib/builder.dart



import 'package:build/build.dart';
import 'package:mobx_codegen/mobx_codegen.dart';
import 'package:source_gen/source_gen.dart';

Builder storeGenerator(BuilderOptions options) =>
    PartBuilder([StoreGenerator()], '.mobx.dart');
```



Disclaimer: I haven't tested it. But it _should_ work, or at least be on the right track.

**Comment by shyndman** [comment]
Cool. Yeah, that looks promising. I'm game for doing it, if @pavanpodila gives the OK.

**Comment by pavanpodila** [comment]
I am still trying to wrap my head around this discussion with freezed. What is the benefit here ?

**Comment by shyndman** [comment]
freezed's builder is able to reference types created by other builders, in large part because it doesn't use `source_gen`. We could do the same, but it (probably) means that we need to introduce our own extension (`.mobx.dart`). If this is the case, it would be a breaking change.

The alternative being, we reintroduce the type name finder that pulls names from the AST, **as a fallback** to the current system that reads from the element model.

**Comment by pavanpodila** [comment]
Ah, thanks for the explanation @shyndman! Can we use the type-discovery feature of freezed directly?. This would avoid the breaking change of using `*.mobx.dart`. Not sure if @rrousselGit would be interested in moving the type-discovery to its own package ?

**Comment by shyndman** [comment]
No, it isn't anything `freezed` is doing specifically. The problem is that `source_gen` is unable to see other types generated via `source_gen` (like those of super_enum, built_value, etc).

**Comment by pavanpodila** [comment]
Got it. Let's make a PR with sufficient tests to tackle all possible situations and then we can introduce the breaking change ? This would more like a `v1.5`

**Comment by shyndman** [comment]
It would have to be a 2.0.0, now that we're past the 1.0.0 mark. Otherwise people matching on `^1.*.*` will be broken after an upgrade.

**Comment by shyndman** [comment]
Unless I'm mistaken about the need for an extension, but I'm reasonably sure it's necessary, otherwise we'd be colliding with other `source_gen` part builders.

**Comment by pavanpodila** [comment]
Ok that makes sense. We will need a migration guide too. MobX is becoming an adult now ;-)

**Comment by rrousselGit** [comment]
A first step could be to just document how to modify the build.yaml file to explain how to do that

There's no change required in mobx to make it work in theory

**Comment by rrousselGit** [comment]
Then, the breaking change could be delayed or made smoother because there's an actual solution.

**Comment by shyndman** [comment]
My understanding is that the parts aren't visible to the analyzer until they've been moved out of the cache, and into their declared "g.dart" locations by the combining builder.

...but, I could be wrong. I'll give it a shot.

**Comment by talamaska** [comment]
@rrousselGit tried the alternative build.yaml and the other stuff, it's not generating anything and no errors in the console. Any help will be appreciated. I'm using moor models and it seems like this was working before, like before 6 months or something. Should I change something in pubspec as well?

**Comment by Ali1Ammar** [comment]
any update on that

**Comment by Ali1Ammar** [comment]
with moor i use [this](https://github.com/simolus3/moor/issues/89#issuecomment-583125248) it working 
that my final build.yaml
```

targets:
  $default:
    builders:
      moor_generator:
        enabled: false
      moor_generator|moor_generator_not_shared:
        enabled: true
      mobx_codegen|mobx_generator:
        enabled: false
      auto_route_generator:autoRouteGenerator:
        enabled: false
      copy_with_extension_gen:
        enabled: false
      json_serializable:
        enabled: false
  run_built_value:
    # run the $default target first, to ensure moor_generator is done
    dependencies: ['learn']
    builders:
      # disable all builders (that are on by default), except for built_value
      moor_generator:
        enabled: false
      moor_generator|preparing_builder:
        enabled: false
```
also we must use .moor insted of .g in moor file.
